enhancements/testing/ephemeral-agent-installer-e2e-tests.md

```diff
@@ -0,0 +1,373 @@
+worked on in the  openshift-agent-team/fleeting repository but this may change in the future.
+
+#### Deployment
+
+Deploying the agent-based installer will be heavily linked to the environment set up. This
+logic will land in dev-scripts alongside the pre environment setup steps.
+
+#### Tests
+
+For the actual test scenarios themselves we should use Ginkgo/Gomega. This is 
+similar to the rest of Openshift. Well most of (if not all) the product is written in Golang
+so it makes sense to use a Golang framework for these.
+
+There is a repository created as a preliminary place to put these. Plan to use a similar
+structure to openshift-tests in this repository.
+
+We will be including existing e2e tests that have been created to validate the OCP cluster
+installed by the new agent based installer.
+
+#### Teardown
+
+Currently there is a `make clean` flow for starting new locally. We will need to add similar
+functionally for the flows that will be used for the agent based installer
+
+### Proposed Refactored Structure
+
+The following is describing general pattern of how to organize these things in the dev-scripts 
+repository. This will be added alongside the existing code.
+
+**Folder Structure:**
+
+```markdown
+playbooks/
+  example-e2e-workflow.yml
+roles/
+    common/
+        tasks/
+        handlers/
+        library/
+        files/
+        templates/
+        vars/
+        defaults/
+        meta/
+    example_role/
+        tasks/
+        defaults/
+        meta/
+        templates/
+        vars/
+```
+
+This is pretty open ended and meant to be. I just believe that focusing on making reusable roles 
+here for the steps is key. With this structure the aim is to be able to easily moving our roles into
+[Ansible Galaxy Modules](https://docs.ansible.com/ansible/latest/galaxy/dev_guide.html) if there is 
+desire for other projects and teams to use them in the future.
+
+**Playbook Structure:**
+
+Use the `pre`, `test`, and `post` pattern in Openshift CI locally in the playbooks and include 
+the playbooks. This will create a structure that will mirror the Openshift CI Step Registry
+and will provide clarity on what code is being run in each step.
+
+```markdown
+# File structure:
+playbooks/
+  example-e2e-workflow.yml
+  example-e2e-workflow_pre.yml
+  example-e2e-workflow_test.yml
+  example-e2e-workflow_post.yml
+
+# example-e2e-workflow.yml
+- hosts: localhost
+
+- name: Run e2e-workflow Pre steps
+  import_playbook: example-e2e-workflow_pre.yml
+
+- name: Run e2e-workflow Test steps
+  import_playbook: example-e2e-workflow_test.yml
+
+- name: Run e2e-workflow Post steps
+  import_playbook: example-e2e-workflow_test.yml
+```
+
+**Parameter Passing and Defaults for Openshift CI**
+
+In Openshift CI the way we pass parameters is through the use of environment variables. 
+This means all of  these Ansible roles will need to have defaults and be aware of those 
+environment variables.
```

Comment:
`dev-scripts` offers an already quick baked-in integration via the the `DEVSCRIPTS_CONFIG` var (see for example [1]). This makes it easy and convenient configuring any job, in additional to the default e2e step default configuration (see [2]). So, as long as the new ansible playbook will be capable to consume the required environment variables then the integration could be performed in a quite direct way.

[1] https://github.com/openshift/release/blob/6ceedc2963da1362f693a4cf7e0ffb8b62a451a6/ci-operator/step-registry/baremetalds/e2e/ovn/ipv6/baremetalds-e2e-ovn-ipv6-workflow.yaml#L5
[2] https://github.com/openshift/release/blob/6ceedc2963da1362f693a4cf7e0ffb8b62a451a6/ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-commands.sh#L99-L104

Reply:
This is a pattern that I've copied and reused in different parts of Openshift CI. It does have a limitation in causing large enumerations of settings which is why I added a couple of other steps in the Assisted folder. But overall it is a pattern I really like. :)

Reply:
I'm not sure if I'm missing something, but the content of `DEVSCRIPTS_CONFIG` is converted to a number of env var. Do you maybe mean to group particular configurations under a single var?

Reply:
This question got lots in the shuffle. I meant that the Ansible playbooks need to use the same environment variables as `DEVSCRIPTS_CONFIG` and `ASSISTED_CONFIG` where it makes sense. [1]

[1]: https://github.com/openshift/assisted-service/blob/master/deploy/operator/ztp/assisted-installer-crds-playbook.yaml#L7-L31